Sources/NIOCore/NIOScheduledCallback.swift

```diff
@@ -0,0 +1,107 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2024 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+/// A type that handles callbacks scheduled with `EventLoop.scheduleCallback(at:handler:)`.
+///
+/// - Seealso: `EventLoop.scheduleCallback(at:handler:)`.
+public protocol NIOScheduledCallbackHandler {
+    /// This function is called at the scheduled time, unless the scheduled callback is cancelled.
+    ///
+    /// - Parameter eventLoop: The event loop on which the callback was scheduled.
+    func handleScheduledCallback(eventLoop: some EventLoop)
```

Comment:
Should there be an optional method to implement when the scheduled callback is cancelled?

Reply:
Not sure really. I think we'd need to see how people use this a bit more to form an opinion on whether it's useful.

Also, if we were to provide it, would it be cleaner to add a non-optional requirement with a default empty implementation?

Reply:
non-optional requirement + default implementation is Swift's equivalent of optional requirement.

FWIW I think having an onCancel callback is useful. It allows you to centralise the state keeping into a single place.